src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SQLite.Interop;
+using Microsoft.CodeAnalysis.Storage;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Lock protecting the write queues and <see cref="_flushAllTask"/>.
+        /// </summary>
+        private readonly object _writeQueueGate = new object();
+
+        /// <summary>
+        /// Task kicked off to actually do the work of flushing all data to the DB.
+        /// </summary>
+        private Task _flushAllTask;
+
+        private void AddWriteTask<TKey>(MultiDictionary<TKey, Action<SqlConnection>> queue, TKey key, Action<SqlConnection> action)
+        {
+            lock (_writeQueueGate)
+            {
+                queue.Add(key, action);
+
+                // If we don't have an outstanding request to write the queue to the DB
+                // then create one to run a short while from now.  If there is an outstanding
+                // request, then it will see this write request when it runs.
+                if (_flushAllTask == null)
+                {
+                    _flushAllTask =
+                        Task.Delay(500, _shutdownTokenSource.Token)
```

Comment:
How was this chosen?  Should we have a reg key to tune it?

At a minimum, should it use `TimeSpan.FromMilliseconds` or something to indicate the unit?